### DISCUSSION TITLE
iTwin/iot-demo errors

### DISCUSSION THREAD

**Discussion by dv-anfimaude** [iTwin/community#37] [Q&A]
I've tried to configured Visualizer to run locally as per this tutorial https://github.com/iTwin/iot-demo/blob/master/IoTVisualizer/README.md
but faced with following error during app start

BE_SQLITE_ERROR: ECClass 'IoTDevice.devicepointobservesspatialelement' does not exist or could not be loaded.
    at DbQueryError.throwIfError (http://localhost:3000/static/js/vendors~main.chunk.js:267382:13)
    at ECSqlReader.runWithRetry (http://localhost:3000/static/js/vendors~main.chunk.js:268949:67)
    at async ECSqlReader.readRows (http://localhost:3000/static/js/vendors~main.chunk.js:268923:18)
    at async ECSqlReader.fetchRows (http://localhost:3000/static/js/vendors~main.chunk.js:268989:23)
    at async ECSqlReader.step (http://localhost:3000/static/js/vendors~main.chunk.js:269002:7)
    at async CheckpointConnection.query (http://localhost:3000/static/js/vendors~main.chunk.js:308957:12)
    at async push../src/IoTExtension/clients/ITwinLink.ts.ITwinLink.fetchDevices (http://localhost:3000/static/js/main.chunk.js:2759:22)
This is an error response from
https://api.bentley.com/imodel/rpc//mode/1/context/265d3ac7-5f8e-48fa-91f0-577e8f897efc/imodel/b440b13e-5127-45e7-b71c-d0b07c82df23/changeset/0/IModelReadRpcInterface-3.4.0-queryRows
endpoint
There is also other similar error :
ECClass 'RevitDynamic.level' does not exist or could not be loaded
I'am using Bentley Retail building sample template model.

Thank in advance!

**Comment by SheetalNair** [comment]
@dv-anfimaude , This project was tested on an iModel with some additional schema provisions. There are few workarounds to get past this error. If you could let us know about your exact requirements, we could guide you in the right direction. 
Do you just want to check out how this sample project works OR do you have a specific IOT problem statement that you are trying to solve?

**Reply by dv-anfimaude** [reply]
@SheetalNair For now - just want to see how this sample works. Could you please provide this iModel with additional schema?

**Comment by SheetalNair** [comment]
@dv-anfimaude, We have used a demo schema to provision our imodel and added some devices for testing, and the fetchDevices() in iTwinLinks.ts file  is basically retriving that information from the iModel. As you do not have that provisioned iModel, you can replace the existing code and make the fetchDevices method return a hardcoded array of objects which looks like this
[{
label: "Concourse-Room1-T1",
id: "0xb3000000003e",
origin: {
"x": 47.471548240841656,
"y": 29.16899822181937,
"z": -2.8318770791543475
},
type: "TC-Room-Temperature-Sensor",
phenomenon: "Temperature",
iotId: "iothub:device:Concourse-Room1-T1",
observedElement: "0x200000010db"
}]

Where,
id: element id of the sensor in the imodel
label: name of the sensor
origin: Point3D coordinates denoting the location of the sensor in the imodel
type: Device type
iotId: The id of the sensor shown in AWS
phenomenon: phenomenon the sensor measures
observedElement: element id of the area monitored by the sensor in the iModel

Note: If you do not know what should be there for id,observedElement or origin value, you can use the same values I have provided in the above sample object and use a Bentley metrostation model and select Concourse level from the widget. The iotId property can be the id of your AWS thing / Azure iot device if you have an existing device in AWS/Azure Iot systems.
